Add multipers

DavidLapous opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| Go              | `@conda-forge/help-go`        |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Zulip chat][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://conda-forge.zulipchat.com

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/multipers/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/multipers/meta.yaml**:

 * ❌ The home item is expected in the about section.
 * ❌ The recipe must have some tests.
 * ❌ Recipe maintainer "David Loiseaux" does not exist
 * ❌ ``requirements: host: numpy>=2`` must contain a space between the name and the pin, i.e. ``numpy >=2``
 * ❌ ``requirements: host: cython>=3`` must contain a space between the name and the pin, i.e. ``cython >=3``

For **recipes/multipers/meta.yaml**:

 * ℹ️ The `boost` output has been superseded by `libboost-python-devel` (as of 1.82), which now comes with a run-export (on `libboost-python`) as well.
 * ℹ️ The `boost-cpp` output has been superseded by `libboost-devel` (as of 1.82), which now comes with a run-export (on `libboost`) as well. In case you only needed the boost headers, you should use `libboost-headers`.
 * ℹ️ Recipes should usually depend on `matplotlib-base` as opposed to `matplotlib` so that runtime environments do not require large packages like `qt`.
 * ℹ️ `noarch: python` recipes should usually follow the syntax in our [documentation](https://conda-forge.org/docs/maintainer/knowledge_base/#noarch-python) for specifying the Python version.
   - For the `host` section of recipe, you should usually use `python {{ python_min }}` for the `python` entry.
   - For the `run` section of recipe, you should usually use `python >={{ python_min }}` for the `python` entry.
   - For the `test.requires` section of recipe, you should usually use `python {{ python_min }}` for the `python` entry.
   - If the package requires a newer Python version than the currently supported minimum version on `conda-forge`, you can override the `python_min` variable by adding a Jinja2 `set` statement at the top of your recipe (or using an equivalent `context` variable for v1 recipes).
 * ℹ️ Jinja2 variable references are suggested to take a ``{{<one space><variable name><one space>}}`` form. See lines [22].


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13057133735](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13057133735). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/multipers/meta.yaml```) and found it was in an excellent condition.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/multipers/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/multipers/meta.yaml**:

 * ❌ Non noarch packages should have python requirement without any version constraints.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13063137412](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13063137412). Examine the logs at this URL for more detail.</sub>

DavidLapous commented:
Dear @conda-forge/help-python-c 

I think this is almost ready, and it seems to build properly on linux + macos. 

I have some questions:
 - [disabling windows release] `multipers` is not yet compatible with windows. I did a
   ```yaml
   build:
     skip: true  # [win]
   ```
   I hope this is the correct way to do this (I didn't find this part in the doc)
 - [python version] for `host`, `build` and `run`, I added a 
   ```
    - python >=3.10, <3.13
   ```
   since `multipers` requires `>=3.10` and a dependency (`gudhi`) requires `<3.13`.
   The linter complains with 
   ```
   Non noarch packages should have python requirement without any version constraints.
   ```
   but I don't really understand this --- what I added are precisely version constraints.
 - [maintainers] it had an error in the first version, and it is fixed now (although the bot didn't update this).
   
Also, I have the hand on the source repo, so if changing something there can help this process, feel free to ask for a change there. I plan to make a new release before the end of this process anyway.

Thanks for your help and review !

synapticarbors commented on recipes/multipers/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+    - {{ compiler("c") }}
+    - {{ stdlib('c') }}
+    - {{ compiler('cxx') }}
+    - python >={{ python_min }},<{{ python_max }}
```
```suggestion
    - python
```

synapticarbors commented on recipes/multipers/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+    - numpy >=2
+    - cython >=3  
+  host:
+    - python >={{ python_min }},<{{ python_max }}
```
```suggestion
    - python
```

synapticarbors commented on recipes/multipers/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+    - libboost-python-devel
+    - libboost-devel
+  run:
+    - python >={{ python_min }},<{{ python_max }}
```
```suggestion
    - python
```

synapticarbors commented on recipes/multipers/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+    - pot
+
+
+test:
```
You should probably just have:

```
test:
   imports:
      - multipers
   requires:
      - pip
   commands:
      - pip check
```

```

DavidLapous replied:
For reasons that I don't understand yet, the `import multipers` doesn't work in the root directory (since python goes first to local files for the import), so I have to `cd` first.

synapticarbors replied:
That's strange. My understanding of our tests is that we don't pull anything from the sdist unless explicitly asked for and we install the package into a conda env based on the local build artifact. Basically you should just get the  conda package, the `run` dependencies, and anything that appears under `requires`.

DavidLapous replied:
It seems to work indeed; thanks !

synapticarbors commented on recipes/multipers/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
You can remove this file since it's in the src distribution on `pypi` (assuming you switch to that as the source)

synapticarbors commented on recipes/multipers/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+    {{ PYTHON }} setup.py build_ext -j${CPU_COUNT}
+    {{ PYTHON }} -m pip install . --no-deps --no-build-isolation -vv
+  number: {{ build }}
+  skip: true  # [win]
```
```suggestion
  skip: true  # [win or py<310]
```

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/multipers/meta.yaml```) and found it was in an excellent condition.

DavidLapous commented on recipes/multipers/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+
+build:
+  script: |
+    {{ PYTHON }} setup.py build_ext -j${CPU_COUNT}
```
The only advantage of this line is multithreaded build. `pip install` only does single-core compilations

DavidLapous commented on recipes/multipers/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
```
One of the reasons that I'm building it here is that I have linked libraries, which are included in the pip package (after an `auditwheel`) which is not required here since they can just be added as dependencies.

My current build system for pip is also a conda environment (`cxx-compiler` included) which is (to my knowledge) not as clean as what conda-forge. In particular, as `c_stdlib_versions` can be better fixed with conda-forge, this should allow older systems to install this library.  

If these are not  "strong reasons" I can go back to the pypi package.

DavidLapous replied:
I think I didn't understand your point:
If the sources are set to pypi, then will conda use directly the compiled version or the sources (`sdist`) ? If it's the sources, then my points are not relevant, and I'll switch to pypi sources.

synapticarbors replied:
If you look in https://github.com/conda-forge/staged-recipes/blob/main/recipes/example/meta.yaml, you'll see that the canonical url we recommend is:

`url: https://pypi.org/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz`

So it's being built from the sdist.

DavidLapous replied:
As mentioned below ([here](https://github.com/conda-forge/staged-recipes/pull/28987#issuecomment-2634065279)), I had to use a `url` on a github release since my github action produced corrupted sources on pip.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint --conda-forge .` from the recipe directory. You can also examine the [workflow logs](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13138168490) for more detail.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13138168490](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13138168490). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/multipers/meta.yaml```) and found it was in an excellent condition.

DavidLapous commented:
I think I know the culprit for the windows version bug; it seems that `multipers` has troubles compiling with old versions of `boost` (here `libboost-devel=1.84`) on windows, which I think are installed here since `gudhi` is there. 

One possibility is simply to remove `gudhi` from the dependency, or disable the windows build until the next version of `gudhi` is released. Since windows support on `multipers` is new, I think I will disable the windows build until it works.

DavidLapous commented:
Another question that I forgot to ask : 

During the compilation there is a `tbb` link. Should I do something to pin the `tbb` `run` version to the `build` version ?
